## **Revision**

In [89]:
# import necessary libraries

import numpy as np # for linear algebra
import pandas as pd # for creating dataframes

from statsmodels.formula.api import ols # for Ordinary Least Squares method

In [90]:
# import the dataset
# you create a dataframe called anyhow you like
# read_csv saves the csv data to a dataframe
# .head(n) function displays top n rows of the dataframe
# .head() displays top 5 rows by default

AppleS_for_SAleeee3 = pd.read_csv('sleep75.csv')
AppleS_for_SAleeee3.head()

,age,black,case,clerical,construc,educ,earns74,gdhlth,inlf,leis1,...,spwrk75,totwrk,union,worknrm,workscnd,exper,yngkid,yrsmarr,hrwage,agesq
0,32,0,1,0.0,0.0,12,0,0,1,3529,...,0,3438,0,3438,0,14,0,13,7.070004,1024
1,31,0,2,0.0,0.0,14,9500,1,1,2140,...,0,5020,0,5020,0,11,0,0,1.429999,961
2,44,0,3,0.0,0.0,17,42500,1,1,4595,...,1,2815,0,2815,0,21,0,0,20.529997,1936
3,30,0,4,0.0,0.0,12,42500,1,1,3211,...,1,3786,0,3786,0,12,0,12,9.619998,900
4,64,0,5,0.0,0.0,14,2500,1,1,4052,...,1,2580,0,2580,0,44,0,33,2.750000,4096


In [91]:
# display the list of dataset columns names to copy them as explaining variables/regressors/'x1', 'x2', 'x3', ...

AppleS_for_SAleeee3.columns.tolist()

['age',
 'black',
 'case',
 'clerical',
 'construc',
 'educ',
 'earns74',
 'gdhlth',
 'inlf',
 'leis1',
 'leis2',
 'leis3',
 'smsa',
 'lhrwage',
 'lothinc',
 'male',
 'marr',
 'prot',
 'rlxall',
 'selfe',
 'sleep',
 'slpnaps',
 'south',
 'spsepay',
 'spwrk75',
 'totwrk',
 'union',
 'worknrm',
 'workscnd',
 'exper',
 'yngkid',
 'yrsmarr',
 'hrwage',
 'agesq']

In [92]:
# specify the regression with OLS
# you create an object called anyhow you like
# 'formula'specifies the regression formula
# 'data' specifies the data and must correspond to the dataframe
# regressors ('x1', 'x2', 'x3', ...) in 'formula' must correspond to the dataframe columns in 'data'
# the variable 'y' (depending variable) is written before '~' ('~' is the same as '=' in the equation)
# successful execution of the cell will result in blank output (the model will be created and memorized)

Mgi1m0_iz_da_b3SSt = ols(formula='sleep~educ+age+hrwage', data=AppleS_for_SAleeee3).fit()

In [93]:
# .summary() displays the complete summary of the model

Mgi1m0_iz_da_b3SSt.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sleep   R-squared:                       0.007
Model:                            OLS   Adj. R-squared:                  0.002
Method:                 Least Squares   F-statistic:                     1.297
Date:                Fri, 24 Apr 2026   Prob (F-statistic):              0.275
Time:                        17:49:23   Log-Likelihood:                -3979.6
No. Observations:                 532   AIC:                             7967.
Df Residuals:                     528   BIC:                             7984.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   3298.7773    127.214     25.931      0.000    3048.870    3548.685
educ          -6.5500      7.505     -0.873      0.383     -21.293       8.193
age            1.8227      1.746      1.044      0.297      -1.607       5.253
hrwage        -5.0691      5.311     -0.954      0.340     -15.503       5.365
==============================================================================
Omnibus:                       14.966   Durbin-Watson:                   1.881
Prob(Omnibus):                  0.001   Jarque-Bera (JB):               26.887
Skew:                          -0.149   Prob(JB):                     1.45e-06
Kurtosis:                       4.060   Cond. No.                         287.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [94]:
# displays the regression's parameters/coefficients before 'x1', 'x2', 'x3', ...
# .round(n) rounds the coefficients to n decimal places
# .round(6) seems to be the limit of possible decimal places

Mgi1m0_iz_da_b3SSt.params.round(3)

Intercept    3298.777
educ           -6.550
age             1.823
hrwage         -5.069
dtype: float64

## **New model for prediction for 'Labour' dataset**

In [95]:
all_work_and_no_play_makes_jack_a_dull_boy = pd.read_csv('Labour.csv')
all_work_and_no_play_makes_jack_a_dull_boy.head(5)

,capital,labour,output,wage
0,2.606563,184,9.250759,43.080307
1,1.323237,91,3.664310,27.780016
2,22.093692,426,28.781516,44.467748
3,10.737851,72,4.124642,39.734710
4,1.161365,46,2.890150,34.650709


In [96]:
Regression1_my_creators_didnt_pay_much_effort_to_name_me_in_an_interesting_way = ols(formula='wage~capital+labour+output', data=all_work_and_no_play_makes_jack_a_dull_boy).fit()

In [97]:
Regression1_my_creators_didnt_pay_much_effort_to_name_me_in_an_interesting_way.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   wage   R-squared:                       0.379
Model:                            OLS   Adj. R-squared:                  0.376
Method:                 Least Squares   F-statistic:                     115.0
Date:                Fri, 24 Apr 2026   Prob (F-statistic):           4.03e-58
Time:                        17:49:23   Log-Likelihood:                -2190.7
No. Observations:                 569   AIC:                             4389.
Df Residuals:                     565   BIC:                             4407.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     38.1080      0.518     73.512      0.000      37.090      39.126
capital       -0.2460      0.022    -11.254      0.000      -0.289      -0.203
labour        -0.0360      0.003    -13.446      0.000      -0.041      -0.031
output         0.7193      0.045     16.094      0.000       0.632       0.807
==============================================================================
Omnibus:                      227.675   Durbin-Watson:                   2.029
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2166.422
Skew:                           1.492   Prob(JB):                         0.00
Kurtosis:                      12.082   Cond. No.                         705.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Let's say, there's a person with

labour = 40;

capital = 18;

output = 3.

### We need to predict wage for this person

In [98]:
Regression1_my_creators_didnt_pay_much_effort_to_name_me_in_an_interesting_way.params.round(3)

Intercept    38.108
capital      -0.246
labour       -0.036
output        0.719
dtype: float64

In [99]:
# The predicted wage will be:
38.108 -0.246*18 -0.036*40 +0.719*3

34.397000000000006

## **New model for prediction for 'sleep75' dataset**

In [100]:
import numpy as np
import pandas as pd

from statsmodels.formula.api import ols

# Not to show Warnings
import warnings
warnings.simplefilter(action='ignore', category=Warning)

In [101]:
# import new data
df = pd.read_csv('sleep75.csv')

In [102]:
# specification sleep=b0+b1age+b2male 
# (how age and gender (male) are connected with sleep)
# y~ x1+x2+x3...
my_model = ols(formula = 'sleep~age+male', data = df).fit()

In [103]:
my_model.params.round(3)

Intercept    3146.745
age             3.589
male          -34.803
dtype: float64

**REGRESSION**

sleep = 3146.745 + 3.589\*age-34.803\*male

### *Forecast the amount of sleep for a person who's 20 years old and male*

In [104]:
# Plug in the Xs multiplying by the coeffs
3146.745 + 3.589*20-34.803*1

3183.722

#### **Forecast**: according to our model a person who's 20 years old and male sleeps 3183.722  minutes a week

## **For the dataset wage1 consider a regression log(wage)  exper, exper2, female, married, smsa.**

In [105]:
import numpy as np
import pandas as pd

from statsmodels.formula.api import ols

# Not to show Warnings
import warnings
warnings.simplefilter(action='ignore', category=Warning)

In [106]:
df2 = pd.read_csv('wage1.csv')

In [107]:
mod = ols(formula = 'np.log(wage)~exper+ I(exper**2)+ female + married + smsa', data = df2)
res = mod.fit()

In [108]:
res.params.round(4)

Intercept        1.2504
exper            0.0369
I(exper ** 2)   -0.0008
female          -0.3627
married          0.1438
smsa             0.2726
dtype: float64

#### Consider individuals with the following characteristics: 

|   |       |        |         |      |
|---|-------|--------|---------|------|
|   | exper | female | married | smsa |
| 1 | 5     | 1      | 1       | 1    |
| 2 | 26    | 0      | 0       | 1    |
| 3 | 38    | 1      | 1       | 0    |

#### Evaluate the prediction of the dependent variable for each individual.

**REGRESSION**

log(wage) =  1.250423 + 0.036880exper-0.000772exper^2
-0.362740female+0.143784married+0.272562smsa

In [109]:
# compute log(wage) for person 1
# we need to plug in the values we have (year of experience, gender etc)
log_wage = 1.2504+0.0369*5-0.0008*np.square(5)-0.3627*1+0.1438*1+0.2726*1
log_wage.round(4)

1.4686

In [110]:
# 1.4686 is the natural log of the wage (remember we typed 'np.log(wage)' in the formula as Y), but we need the actual wage
# Thus we need the exponent
# wage = np.exp(1.4686)
np.exp(1.4686)

4.343150472127337

#### **Answer**: The predicted wage for person 1 (5 years of experience, female, married and lives in smsa) is 4.345

#### **The predicted wage for person 2**

In [111]:
np.exp(1.2504+0.0369*26-0.0008*np.square(26)-0.3627*0+0.1438*0+0.2726*1)

6.969893884143942

## **For the dataset Labour consider a regression log(output) on log(capital), log(labour)**

#### Consider a firm with the following characteristics:

|  capital | labour |
|----------|--------|
|  2.970   |  85    |


#### Calculate the predicted output

In [112]:
df3 = pd.read_csv('Labour.csv')

In [113]:
regression3 = ols('np.log(output)~ np.log(capital)+ np.log(labour)', data=df3).fit()

In [114]:
regression3.params.round(3)

Intercept         -1.711
np.log(capital)    0.208
np.log(labour)     0.715
dtype: float64

In [115]:
log_output=-1.711+0.208*np.log(2.970)+0.715*np.log(85)

In [116]:
# The output of the firm is
np.exp(log_output).round(4)

5.4299